ProcessLib/RichardsFlow/RichardsFlowFEM.h

```diff
@@ -193,27 +247,28 @@ class LocalAssemblerData : public RichardsFlowLocalAssemblerInterface
 
         for (unsigned ip = 0; ip < n_integration_points; ++ip)
         {
-            auto const& sm = _shape_matrices[ip];
-
-            double P_int_pt = 0.0;
-            NumLib::shapeFunctionInterpolate(local_x, sm.N, P_int_pt);
-            double const Pc = -P_int_pt;
-            double const Sw = interpolated_Pc.getValue(Pc);
-            double const k_rel = interpolated_Kr.getValue(Sw);
-            double const K_mat_coeff = K * k_rel / mu;
-
-            GlobalDimVectorType velocity =
-                -K_mat_coeff * sm.dNdx * p_nodal_values;
+            double p_int_pt = 0.0;
+            NumLib::shapeFunctionInterpolate(local_x, _ip_data[ip].N, p_int_pt);
+            double const pc_int_pt = -p_int_pt;
+            const double& temperature = _process_data.temperature(t, pos)[0];
+            double const Sw = _process_data.material->getSaturation(
+                material_id, t, pos, p_int_pt, temperature, pc_int_pt);
+            double const k_rel =
+                _process_data.material->getRelativePermeability(
+                    t, pos, p_int_pt, temperature, Sw);
+            auto const mu = _process_data.material->getFluidViscosity(
+                p_int_pt, temperature);
+            GlobalDimVectorType velocity = -permeability * (k_rel / mu) *
+                                           _ip_data[ip].dNdx * p_nodal_values;
             if (_process_data.has_gravity)
             {
-                auto const rho_w = _process_data.water_density(t, pos)[0];
-                auto const body_force =
-                    _process_data.specific_body_force(t, pos);
+                auto const rho_w = _process_data.material->getFluidDensity(
+                    p_int_pt, temperature);
+                auto const body_force = _process_data.specific_body_force;
                 assert(body_force.size() == GlobalDim);
-                auto const b = MathLib::toVector<GlobalDimVectorType>(
-                    body_force, GlobalDim);
-                // here it is assumed that the vector b is directed 'downwards'
-                velocity += K_mat_coeff * rho_w * b;
+                // here it is assumed that the vector body_force is directed
+                // 'downwards'
+                velocity += permeability * (k_rel / mu) * rho_w * body_force;
```

Comment:
For the performance concerned programmers:
 - permeability * k_rel/mu is always used together, so it can be evaluated once.
 - velocity is always multiplied with `permeability * k_rel / mu`, so may be it can be done once after the if-condition.